internal/rbd/nodeserver.go

```diff
@@ -1221,31 +1230,100 @@ func (ns *NodeServer) xfsSupportsReflink() bool {
 	return false
 }
 
+// extractPVNameFromPath extracts PV name from path containing PV name.
+func extractPVNameFromPath(path string) string {
+	str := strings.Split(path, "/pvc-")[1]
+	str = strings.Split(str, "/")[0]
+
+	return "pvc-" + str
+}
+
 // NodeGetVolumeStats returns volume stats.
 func (ns *NodeServer) NodeGetVolumeStats(
 	ctx context.Context,
 	req *csi.NodeGetVolumeStatsRequest,
 ) (*csi.NodeGetVolumeStatsResponse, error) {
-	var err error
-	targetPath := req.GetVolumePath()
-	if targetPath == "" {
-		err = fmt.Errorf("targetpath %v is empty", targetPath)
+	var (
+		imgInfo rbdImageMetadataStash
+		err     error
+	)
+	var metadataPath string
+	volumePath := req.GetStagingTargetPath()
+	if volumePath == "" {
+		if volumePath = req.GetVolumePath(); volumePath == "" {
+			return nil, status.Error(codes.InvalidArgument, "volume path must be provided")
+		}
+		pvName := extractPVNameFromPath(volumePath)
+		metadataPath = ns.StagingPath + pvName + "/globalmount"
+	} else {
+		metadataPath = volumePath
+		volumePath = volumePath + "/" + req.GetVolumeId()
+	}
 
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+	imgInfo, err = lookupRBDImageMetadataStash(metadataPath)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to find image metadata at stagingPath %q: %v", volumePath, err)
+
+		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	stat, err := os.Stat(targetPath)
+	return checkBlockDeviceStatus(ctx, volumePath, &imgInfo, ns.Mounter)
+}
+
+// checkBlockDeviceStatus checks the block device current status.
+func checkBlockDeviceStatus(
+	ctx context.Context,
+	volumePath string,
+	imgInfo *rbdImageMetadataStash,
+	mounter mount.Interface,
+) (*csi.NodeGetVolumeStatsResponse, error) {
+	var volCondition csi.VolumeCondition
+
+	stat, err := os.Stat(volumePath)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, "failed to get stat for targetpath %q: %v", targetPath, err)
+		volCondition.Abnormal = true
+		volCondition.Message = fmt.Sprintf("failed to get stat for volumepath %q: %v", volumePath, err)
+
+		return &csi.NodeGetVolumeStatsResponse{
+			VolumeCondition: &volCondition,
+		}, nil
+	}
+
+	// userspace mounters
```

Comment:
It would be cleaner to have this check in a helper function like `checkHealthNBD(volume, path)`. The function can return an error that can then be used for marking the volume unhealthy.